### ISSUE TITLE
Begin work on the NLG task

### ISSUE THREAD

**Issue by AmateurAcademic** [AmateurAcademic/nlu-engine#9]
# Description
We don't just want a model that can do NLU tasks, we also want it to be able to do voice assistant like NLG. The NLG should work with passing it the utterance, intent, domain, and any important outputs from an API and generate a response. 

## Example
Utterance: Turn off the kitchen lights
Domain: IoT
Intent: turn_off
API response: state of kitchen lights goes from on to off
response: The kitchen lights are now off. 

## Questions and challenges
* I will start by using my own data set for my personal voice assistant as a first version of data. One of the challenges will be getting more data. 
* Another challenge will be covering enough responses from APIs, there are so many with different formats that can provide lots of information. It will be an interesting challenge providing the fine-tuning of the model with enough examples that it can parse out important information and can be robust enough to handle any type of responses given to it by an API. Take for example weather. Collecting enough data and handling it will not be a trivial task.
* What metric will we use for the responses? ROUGE? And how will we judge the quality of the responses?

# DoD
- [x] Gather first version responses data
- [x] Get first version of API responses data
- [x] Create a data set from this
- [x] Come up with a good prompt, it might require giving an example along with a description (similar to entity extraction)
- [x] Create a notebook to run inference for this task and measure performance
- [x] See how well a larger model performs with zero and few shot

**Comment by AmateurAcademic** [comment]
The nexts steps will be to:
* try out the preprocessing for for the nlg data on creating training data, one for the test inference config
* try a non-fine-tuned model on it's abilities with the inference test data set.
* try to train a model on the task, using the small data set: need to create a good loss metric for this, too.
* annotate more data
* then go from there with further fine-tuning and testing

**Comment by AmateurAcademic** [comment]
The CSV I have for the NLG simulation data, is very complex and therefore has lots of errors in formatting for reading the CSV. I will have to manually go through the CSV and replace the commas with semicolons to ensure the columns are being read correctly, then make a minor code modification to read semicolons as delimiters. What a shame, I really wanted to process the file today and start training the NLG stuff in addition to the NLU stuff. I will have to work on this next session, as I am out of time.

**Comment by AmateurAcademic** [comment]
I finally had enough time to fix all of the data entries, so I can continue to work on this issue again.

**Comment by AmateurAcademic** [comment]
I have completed the cleaning of the data, and well as updated the preprocessing script to handle creating a training set using the NLG data. 

The next step is to update the trainer.py script so that we can get ROUGE scores from compute_metrics. 

For now, I will use eval_loss for the best model selection, but I might experiment with combining the scores from the compute_metrics and see if that works better.

**Comment by AmateurAcademic** [comment]
I have updated the trainer.py. Next session, I will test it by training an nlg-only model. It will probably mess up some how and I will spend my entire next session debugging it and hopefully getting it to work. But, hey, I might get lucky, and it might just work. Yeah sure...

**Comment by AmateurAcademic** [comment]
I have finally created a proper small data set of 350 entries. Next step will be to train a new model and see if we need more.

**Comment by AmateurAcademic** [comment]
The training on the NLG dataset was successful. A follow up training including all tasks was performed. Fine-tuning on multiple NLU-NLG tasks did not degrade any of the performance. This shows that this approach works. 

This final test closes this issue.
